### PR TITLE
Upgrade react router to a stable release

### PR DESCRIPTION
**Pull request by ajainarayanan**
This PR is restarting efforts that went in as part of this PR https://github.com/caskdata/cdap/pull/8385

+ Add unit tests for components related to this change.

Components that require unit tests,

```
- EntityListView
- Overview (all entities overview)
- Detailed view (all entities detailed view)
- Header (including ProductDropdown, NamespaceDropdown)
- BreadCrumb,
- 404
```
This is to start with. Will add unit tests for other components in subsequent PR.

Bamboo build: https://builds.cask.co/browse/CDAP-DRC6419-1

### CODE REVIEW

**Comment by tonybach** [comment]
Seems like there's a problem with the detail view of an application. Navigating directly to the URL doesn't work either. Here's the error:

screen shot 2017-04-18 at 10 21 18 am

File: cdap-ui/app/cdap/api/__mocks__/app.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+MyAppApi.batchStatus = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return this.__batchStatusStats;
+  });
+  return subject;
+};
+export {MyAppApi};
+
+// const list = () => {};
```
Comment (tonybach) [inline cdap-ui/app/cdap/api/__mocks__/app.js]: Do you want to keep these commented-out code?

File: cdap-ui/app/cdap/api/__mocks__/app.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  },
+  __batchStatusStats: []
+};
+MyAppApi.__setApp = function(app) {
+  this.__app = app;
+};
+MyAppApi.get = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return this.__app;
```
Comment (tonybach) [inline cdap-ui/app/cdap/api/__mocks__/app.js]: `subject.onNext(this.__app)`?

File: cdap-ui/app/cdap/api/__mocks__/dataset.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+MyDatasetApi.list = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    subject.onNext(this.__list);
+  });
+  return subject;
+};
+
+export {MyDatasetApi};
```
Comment (tonybach) [inline cdap-ui/app/cdap/api/__mocks__/dataset.js]: Not sure why you're doing `export {}` in some places, `export default` in some others, as well as `module.exports = {}`.

File: cdap-ui/app/cdap/components/EntityListView/NoEntitiesMessage/__tests__/NoEntitiesMessage.test.js [hunk 4]
```diff
@@ -27,6 +27,9 @@ let noEntitiesMessage;
 let filtersAreApplied = jest.fn();
 
 describe('NoEntitiesMessage Unit tests - ', () => {
+  beforeEach(() => {
+    filtersAreApplied = jest.fn();
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/EntityListView/NoEntitiesMessage/__tests__/NoEntitiesMessage.test.js]: I think this is unnecessary as we already have the function expression on line 27. Let me know if I'm missing something.

**Comment by ajainarayanan** [comment]
@tonybach Updated.

File: cdap-ui/app/cdap/api/__mocks__/dataset.js (hunk 3 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/api/__mocks__/dataset.js]: @tonybach the original modules some have a default export and some have named exports. If the mocks that we write have to replace them then they need to do the same thing. Making exports of api/*.js to be consistent can come in another PR as that will have more changes across the code.

File: cdap-ui/app/cdap/api/__mocks__/app.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  },
+  __batchStatusStats: []
+};
+MyAppApi.__setApp = function(app) {
+  this.__app = app;
+};
+MyAppApi.get = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return subject.onNext(this.__app);
```
Comment (tonybach) [inline cdap-ui/app/cdap/api/__mocks__/app.js]: is it `return subject.onNext(this.__app)` or just ` subject.onNext(this.__app)`?

File: cdap-ui/app/cdap/api/__mocks__/app.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  });
+  return subject;
+};
+
+MyAppApi.batchStatus = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return this.__batchStatusStats;
```
Comment (tonybach) [inline cdap-ui/app/cdap/api/__mocks__/app.js]: ` subject.onNext(this.__batchStatusStats)` here too?

File: cdap-ui/app/cdap/api/__mocks__/app.js (hunk 5 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/api/__mocks__/app.js]: Not sure if that would make any difference but will change though.

File: cdap-ui/app/cdap/components/Overview/index.js [hunk 7]
```diff
@@ -258,5 +262,5 @@ export default class Overview extends Component {
 }
 
 Overview.propTypes = {
-  onCloseAndRefresh: PropTypes.func
+  onCloseAndRefresh: PropTypes.func,
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Overview/index.js]: nit: extra comma not needed

File: cdap-ui/app/cdap/components/Overview/StreamOverview/index.js [hunk 8]
```diff
@@ -168,5 +167,5 @@ StreamOverview.propTypes = {
   toggleOverview: PropTypes.bool,
   entity: PropTypes.object,
   onClose: PropTypes.func,
-  onCloseAndRefresh: PropTypes.func
+  onCloseAndRefresh: PropTypes.func,
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/Overview/StreamOverview/index.js]: here as well

**Comment by elfenheart** [comment]
Minor codestyle comments, but otherwise 👍  LGTM. 

Everything seems to work as well, so let's get this merged!

**Comment by tonybach** [comment]
LGTM 👍

**Comment by bdmogal** [comment]
Let's be careful about such PRs in future. For now, since it is in the beginning of the release, and this has been worked on and tested for a while, I think it is ok to merge this.

**Pull request merged**